packages/kotti-ui/source/props.ts

```diff
@@ -0,0 +1,198 @@
+				for (const type of typeSet) types.add(type)
+
+			if (DEBUG_WALK_SHAPE) console.groupEnd()
+		} else {
+			if (DEBUG_WALK_SHAPE) console.log(`walkTypes: pushing “${typeName}”`)
+			types.add(typeName)
+		}
+
+		shapePointer = shapePointer._def.innerType
+	}
+
+	return types
+	/* eslint-enable no-console */
+}
+
+/**
+ * These types do not influence the generation of Vue’s `propName.type` Arrays and are therefore discarded
+ */
+const ignoredZodTypes = new Set(['ZodDefault', 'ZodNullable', 'ZodOptional'])
+
+/**
+ * This maps the internal zod name of a type to the the constructor that Vue expects in `propName.type`
+ */
+const zodToVueTypes = new Map<
+	string,
+	Exclude<PropType<unknown>, Array<unknown>>
+>([
+	['ZodArray', Array],
+	['ZodBoolean', Boolean],
+	['ZodFunction', Function],
+	['ZodNativeEnum', String], // educated guess, can be fixed if the need for non-string enums arises
+	['ZodNumber', Number],
+	['ZodObject', Object],
+	['ZodString', String],
+])
+
+/**
+ * @summary Easily define & validate vue props, improve error messages drastically
+ *
+ * @description
+ * This augments/replaces Vue’s props entirely, including the need to add prop.validator
+ * by using zod and therefore actually prints errors to the console that can help developers
+ * rather than just say "hey you failed the validator, good luck!" as vue would do by default.
+ *
+ * ## Known limitations:
+ *
+ * 1. Deeply defined defaults are only applied when manually `schema.safeParse`-ing the data (only top-level defaults are possible to transform to Vue)
+ * 2. While this function also automatically types the return types via as PropType<> (which in theory
+ *   should mean that annotating the prop types explicitly is no longer necessary) in `@vue/composition-api`
+ *   (`0.6.1` and `1.1.5`), `PropType<>` appears to be bugged.
+ *   For example KtBreadcrumb’s props get inferred to be any
+ *   So, unfortunately, this may only help in the future, once the Vue team fixes this
+ * 3. `z.function().default()` does not accept `undefined` while parsing the schema manually
+ *   without an explicit `.optional()` {@link https://github.com/colinhacks/zod/issues/647}
+ *   However, it can still be used to generate the Vue defaults, as Vue doesn’t call validators with `undefined`
+ *
+ * @example
+ * // KtUserMenu.vue
+ * export default defineComponent<KottiUserMenu.PropsInternal>({
+ *   name: 'KtUserMenu',
+ *   props: makeProps(KottiUserMenu.propsSchema)
+ * })
+ *
+ * // types.ts
+ * export namespace KottiUserMenu {
+ *   const propsSchema = z.object({
+ *     // ...
+ *   })
+ *   type Props = z.input<typeof propsSchema>
+ *   type PropsInternal = z.output<typeof propsSchema>
+ * }
+ */
+export const makeProps = <SCHEMA extends z.ZodObject<z.ZodRawShape>>(
+	schema: SCHEMA,
+): {
+	[KEY in keyof SCHEMA['shape']]: Omit<PropOptions, 'required' | 'type'> & {
+		required: z.output<SCHEMA>[KEY] extends undefined ? false : true
+		type: PropType<z.output<SCHEMA>[KEY]>
```

Comment:
same ☝🏼

Reply:
Actually, as it turns out the `required` should use `z.input` and the `type` should use `z.output`.